# -------------------------------------------------------------------------
# Description:
# This script processes and analyzes orthogroup data generated by OrthoFinder. 
# It includes the following steps:
# 1. Reads and reformats data for per-species orthogroup statistics.
# 2. Generates bar plots to visualize the counts and percentages of genes in 
#    orthogroups and species-specific orthogroups.
# 3. Creates an UpSet plot to display the presence/absence of orthogroups 
#    across genomes.
# 4. Identifies orthogroups that are:
#    - Present in all species (core orthogroups).
#    - Specific to a single species (species-specific orthogroups).
# 5. Calculates the total number of core and specific genes for a target genome.
#
# Key Functions:
# - Integrates ggplot2 for bar plot visualization and ComplexUpset for 
#   advanced UpSet plots.
# - Facilitates the extraction of orthogroup statistics to interpret genome 
#   evolution and relationships.
# - Outputs results as publication-ready plots in PDF format.
#
# Required Libraries:
# - tidyverse
# - data.table
# - ComplexUpset
# - cowplot
#
# Inputs:
# - OrthoFinder output files:
#   - "Statistics_PerSpecies.tsv": Contains per-species statistics.
#   - "Orthogroups.GeneCount.tsv": Contains orthogroup gene counts.
#
# Outputs:
# - "Plots/orthogroup_plot.pdf": Bar plot showing counts of orthogroups and genes.
# - "Plots/orthogroup_percent_plot.pdf": Bar plot showing percentages of genes 
#   in orthogroups.
# - "Plots/one-to-one_orthogroups_plot.complexupset.pdf": UpSet plot showing 
#   orthogroup presence/absence across genomes.
#
# Notes:
# - Ensure the file paths for OrthoFinder outputs are correctly specified.
# - Adjust genome names in the script if needed to match the dataset.
# -------------------------------------------------------------------------


library(tidyverse)
library(data.table)


dat <- fread("/data/users/UserID/assembly-annotation-course/output/maker/final/genespace/orthofinder/Results_Nov01/Comparative_Genomics_Statistics/Statistics_PerSpecies.tsv", header = T, fill = TRUE)
genomes <- names(dat)[names(dat) != "V1"]

dat <- dat %>% pivot_longer(cols = -V1, names_to = "species", values_to = "perc")
ortho_ratio <- dat %>%
    filter(V1 %in% c(
        "Number of genes", "Number of genes in orthogroups", "Number of unassigned genes",
        "Number of orthogroups containing species", "Number of species-specific orthogroups", "Number of genes in species-specific orthogroups"
    ))

ortho_percent <- dat %>%
    filter(V1 %in% c(
        "Percentage of genes in orthogroups", "Percentage of unassigned genes", "Percentage of orthogroups containing species",
        "Percentage of genes in species-specific orthogroups"
    ))

p <- ggplot(ortho_ratio, aes(x = V1, y = perc, fill = species)) +
    geom_col(position = "dodge") +
    cowplot::theme_cowplot() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Count")
ggsave("Plots/orthogroup_plot.pdf")


p <- ggplot(ortho_percent, aes(x = V1, y = as.numeric(perc), fill = species)) +
    geom_col(position = "dodge") +
    ylim(c(0, 100)) +
    cowplot::theme_cowplot() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(y = "Count")
ggsave("Plots/orthogroup_percent_plot.pdf")


library(UpSetR)


orthogroups <- fread("/data/users/UserID/assembly-annotation-course/output/maker/final/genespace/orthofinder/Results_Nov01/Orthogroups/Orthogroups.GeneCount.tsv")
orthogroups <- orthogroups %>%
    select(-Total)
ogroups_presence_absence <- orthogroups
rownames(ogroups_presence_absence) <- ogroups_presence_absence$Orthogroup

# convert the gene counts to presence/absence
ogroups_presence_absence[ogroups_presence_absence > 0] <- 1
ogroups_presence_absence$Orthogroup <- rownames(ogroups_presence_absence)

str(ogroups_presence_absence)

ogroups_presence_absence <- ogroups_presence_absence %>%
    rowwise() %>%
    mutate(SUM = sum(c_across(!ends_with("Orthogroup"))))



ogroups_presence_absence <- data.frame(ogroups_presence_absence)
ogroups_presence_absence[genomes] <- ogroups_presence_absence[genomes] == 1


# use ComplexUpset package to make an upset plot with a subset of the data
library(ComplexUpset)

pdf("Plots/one-to-one_orthogroups_plot.complexupset.pdf", height = 5, width = 10, useDingbats = FALSE)

upset(ogroups_presence_absence, genomes, name = "genre", width_ratio = 0.1, wrap = TRUE, set_sizes = FALSE)
dev.off()

# Orthogroups present in all species
og_all_species <- ogroups_presence_absence %>%
    filter(SUM == length(genomes)) %>%
    select(Orthogroup) %>%
    pull()

# Orthogroups present in only genome1
og_one_species <- ogroups_presence_absence %>%
    filter(SUM == 1) %>%
    filter(genome1 == TRUE) %>%
    select(Orthogroup) %>%
    pull()

# number of CORE genes in genome1
genes_in_og_all_species <- orthogroups %>%
    filter(Orthogroup %in% og_all_species) %>%
    select(genome1) %>%
    sum()

# number of SPECIFIC genes in genome1
genes_in_og_one_species <- orthogroups %>%
    filter(Orthogroup %in% og_one_species) %>%
    select(genome1) %>%
    sum()

# The names of the genes in the orthogroups in Orthogroups.tsv
